segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentMapper.java

```diff
@@ -180,14 +177,88 @@ public StreamSegmentMapper(ContainerMetadata containerMetadata, OperationLog dur
 
         String transactionName = StreamSegmentNameUtils.getTransactionNameFromId(parentStreamSegmentName, transactionId);
         return parentCheck
-                .thenComposeAsync(parentId -> this.storage.create(transactionName, timer.getRemaining()), this.executor)
-                .thenComposeAsync(transInfo -> this.stateStore.put(transInfo.getName(), getState(transInfo, attributes), timer.getRemaining()), this.executor)
+                .thenComposeAsync(parentId -> createSegmentInStorageWithRecovery(transactionName, attributes, timer), this.executor)
                 .thenApply(v -> {
                     LoggerHelpers.traceLeave(log, traceObjectId, "createNewTransactionStreamSegment", traceId, parentStreamSegmentName, transactionName);
                     return transactionName;
                 });
     }
 
+    /**
+     * Attempts to create the given Segment (or Transaction) in Storage, with possible recovery from a previous incomplete
+     * attempt. When this method completes successfully, the Storage will have the Segment created, as well as a State File
+     * with the appropriate contents.
+     * <p>
+     * The recovery part handles these three major cases:
+     * <ul>
+     * <li>Segment exists and has a valid State File: the operation will fail with StreamSegmentExistsException.
+     * <li>Segment exists, has a length of zero, and a missing or invalid State File: the state file will be recreated using
+     * the given attributes and the operation will complete successfully (pending a successful State File creation).
+     * <li>Segment exists, has a non-zero length, and either a valid or invalid/missing State File: the state file will be
+     * recreated (if needed) using the given attributes and the operation will fail with StreamSegmentExistsException.
+     * </ul>
+     *
+     * @param segmentName The name of the Segment/Transaction to create.
+     * @param attributes  The initial Attributes for the Segment, if any.
+     * @param timer       A TimeoutTimer to determine how much time is left to complete the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the Segment has been successfully created.
+     */
+    private CompletableFuture<Void> createSegmentInStorageWithRecovery(String segmentName, Collection<AttributeUpdate> attributes, TimeoutTimer timer) {
+        return FutureHelpers
+                .exceptionallyCompose(
+                        this.storage.create(segmentName, timer.getRemaining()),
+                        ex -> handleStorageCreateException(segmentName, ExceptionHelpers.getRealException(ex), timer))
+                .thenComposeAsync(si ->
```

Comment:
Also minor, but could you rename `si` to something more meaningful? from the signature of `getState`, this is a segment properties object, perhaps `segmentProps` if using properties makes it too long.

Reply:
fixed